File: packages/provide-uterm-go/hub/registry_test.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

package hub

import (
	"errors"
	"testing"
)

func TestRegistryStartsEmpty(t *testing.T) {
	r := NewWorkerRegistry()
	mustEqual(t, r.Len(), 0, "len")
	mustTrue(t, r.Get("nope") == nil, "get missing")
	mustFalse(t, r.Contains("nope"), "contains missing")
	mustEqual(t, len(r.All()), 0, "all")
	mustEqual(t, len(r.Keys()), 0, "keys")
}

func TestRegistryPutGetPopRoundtrip(t *testing.T) {
	r := NewWorkerRegistry()
	st := NewWorkerTermState()
	r.Put("w1", st)
	mustTrue(t, r.Get("w1") == st, "get returns same")
	mustTrue(t, r.Contains("w1"), "contains")
	mustEqual(t, r.Len(), 1, "len")
	mustDeepEqual(t, r.Keys(), []string{"w1"}, "keys")
	mustEqual(t, len(r.All()), 1, "all len")
	mustTrue(t, r.All()[0] == st, "all[0]")
	popped := r.Pop("w1")
	mustTrue(t, popped == st, "pop returns same")
	mustTrue(t, r.Get("w1") == nil, "get after pop")
	mustTrue(t, r.Pop("w1") == nil, "pop absent -> nil")
}

func TestRegistrySetDefaultKeepsExisting(t *testing.T) {
	r := NewWorkerRegistry()
	first := NewWorkerTermState()
	second := NewWorkerTermState()
	mustTrue(t, r.SetDefault("w1", first) == first, "setdefault inserts")
	mustTrue(t, r.SetDefault("w1", second) == first, "setdefault keeps existing")
	mustTrue(t, r.Get("w1") == first, "get is first")
}

func TestRegistryDiscardReturnsTruth(t *testing.T) {
	r := NewWorkerRegistry()
	r.Put("w1", NewWorkerTermState())
	mustTrue(t, r.Discard("w1"), "discard present")
	mustFalse(t, r.Discard("w1"), "discard absent")
}

func TestRegistryRequireRaisesOnMissing(t *testing.T) {
	r := NewWorkerRegistry()
	st := NewWorkerTermState()
	r.Put("w1", st)
	got, err := r.Require("w1")
	mustTrue(t, err == nil && got == st, "require present")
	_, err = r.Require("missing")
	var notFound *ErrWorkerNotFound
	mustTrue(t, errors.As(err, &notFound), "require missing errors")
	mustEqual(t, notFound.WorkerID, "missing", "error names worker id")
	mustTrue(t, notFound.Error() != "", "error string")
}

func TestRegistryKeysSortedSnapshot(t *testing.T) {
	r := NewWorkerRegistry()
	r.Put("b", NewWorkerTermState())
	r.Put("a", NewWorkerTermState())
	r.Put("c", NewWorkerTermState())
	mustDeepEqual(t, r.Keys(), []string{"a", "b", "c"}, "sorted keys")
	mustEqual(t, len(r.All()), 3, "all len")
}

File: packages/provide-uterm-ts/src/hub/registry.test.ts
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

import { describe, expect, it } from "vitest";
import { WorkerRegistry } from "./index.ts";

/** A stand-in for the worker state the registry stores. */
interface State {
  id: string;
}

const a: State = { id: "a" };
const b: State = { id: "b" };

describe("WorkerRegistry lookup", () => {
  it("returns undefined for a worker it does not know", () => {
    expect(new WorkerRegistry<State>().get("nope")).toBeUndefined();
  });

  it("returns the state it was given", () => {
    const registry = new WorkerRegistry<State>();
    registry.put("w1", a);
    expect(registry.get("w1")).toBe(a);
  });

  it("throws from require for an unknown worker", () => {
    expect(() => new WorkerRegistry<State>().require("nope")).toThrow("nope");
  });

  it("returns the state from require when it is known", () => {
    const registry = new WorkerRegistry<State>();
    registry.put("w1", a);
    expect(registry.require("w1")).toBe(a);
  });

  it("reports membership", () => {
    const registry = new WorkerRegistry<State>();
    expect(registry.contains("w1")).toBe(false);
    registry.put("w1", a);
    expect(registry.contains("w1")).toBe(true);
  });
});

describe("WorkerRegistry mutation", () => {
  it("replaces an existing entry on put", () => {
    const registry = new WorkerRegistry<State>();
    registry.put("w1", a);
    registry.put("w1", b);
    expect(registry.get("w1")).toBe(b);
    expect(registry.size).toBe(1);
  });

  it("keeps the existing entry on setDefault", () => {
    const registry = new WorkerRegistry<State>();
    registry.put("w1", a);
    // The point of setDefault: an attach racing a reattach must not discard
    // the live state and hand back a fresh one.
    expect(registry.setDefault("w1", b)).toBe(a);
    expect(registry.get("w1")).toBe(a);
  });

  it("inserts and returns the new entry when setDefault finds nothing", () => {
    const registry = new WorkerRegistry<State>();
    expect(registry.setDefault("w1", a)).toBe(a);
    expect(registry.get("w1")).toBe(a);
  });

  it("returns the removed state from pop", () => {
    const registry = new WorkerRegistry<State>();
    registry.put("w1", a);
    expect(registry.pop("w1")).toBe(a);
    expect(registry.contains("w1")).toBe(false);
  });

  it("returns undefined when popping an unknown worker", () => {
    expect(new WorkerRegistry<State>().pop("nope")).toBeUndefined();
  });

  it("reports whether discard removed anything", () => {
    const registry = new WorkerRegistry<State>();
    registry.put("w1", a);
    expect(registry.discard("w1")).toBe(true);
    expect(registry.discard("w1")).toBe(false);
  });
});

describe("WorkerRegistry snapshots", () => {
  it("counts its entries", () => {
    const registry = new WorkerRegistry<State>();
    expect(registry.size).toBe(0);
    registry.put("w1", a);
    registry.put("w2", b);
    expect(registry.size).toBe(2);
  });

  it("lists ids, states and pairs in insertion order", () => {
    const registry = new WorkerRegistry<State>();
    registry.put("w1", a);
    registry.put("w2", b);
    expect(registry.keys()).toStrictEqual(["w1", "w2"]);
    expect(registry.all()).toStrictEqual([a, b]);
    expect(registry.items()).toStrictEqual([
      ["w1", a],
      ["w2", b],
    ]);
  });

  it("returns snapshots that a later mutation does not disturb", () => {
    // Callers iterate these while the hub mutates the registry, so a live
    // view would throw or skip entries mid-iteration.
    const registry = new WorkerRegistry<State>();
    registry.put("w1", a);
    const keys = registry.keys();
    const states = registry.all();
    const items = registry.items();
    registry.put("w2", b);
    registry.pop("w1");
    expect(keys).toStrictEqual(["w1"]);
    expect(states).toStrictEqual([a]);
    expect(items).toStrictEqual([["w1", a]]);
  });

  it("iterates its ids", () => {
    const registry = new WorkerRegistry<State>();
    registry.put("w1", a);
    registry.put("w2", b);
    expect([...registry]).toStrictEqual(["w1", "w2"]);
  });
});
